### PR TITLE
Add Doxyfile

### PR DESCRIPTION
**Pull request by TobiasFella**
Next steps would be:
- Adding a CI job to build the documentation
- Hosting it somewhere

Where to host it?
I could put it on my server, or we can use github pages

It's almost completely the default template, since I don't really know anything about doxygen - seems to mostly work though

@CarlSchwan do you know anything useful about how doxygen should be configured (any maybe make it less ugly? :smile:)

Current status can be viewed at [quotient.tobiasfella.de](https://quotient.tobiasfella.de)

### CODE REVIEW

**Comment by TobiasFella** [comment]
Some problems I noticed:
- on pages like https://quotient.tobiasfella.de/classRoom.html , it says `Definition at line 82 of file room.cpp` - which is very wrong, the class is defined in the header; the line it points to is the definition of the private class
- In https://quotient.tobiasfella.de/annotated.html, some classes are shown on the top-level of the tree, while others are below the `Quotient` namespace

**Comment by TobiasFella** [comment]
> Some problems I noticed:
> 
>     * on pages like https://quotient.tobiasfella.de/classRoom.html , it says `Definition at line 82 of file room.cpp` - which is very wrong, the class is defined in the header; the line it points to is the definition of the private class
> 
>     * In https://quotient.tobiasfella.de/annotated.html, some classes are shown on the top-level of the tree, while others are below the `Quotient` namespace

Fixed by not building docs for .cpp files

File: Doxyfile [hunk 1]
```diff
@@ -0,0 +1,2764 @@
...
+# (brief and detailed) documentation of class members so that constructors and
+# destructors are listed first. If set to NO the constructors will appear in the
+# respective orders defined by SORT_BRIEF_DOCS and SORT_MEMBER_DOCS.
+# Note: If SORT_BRIEF_DOCS is set to NO this option is ignored for sorting brief
+# member documentation.
+# Note: If SORT_MEMBER_DOCS is set to NO this option is ignored for sorting
+# detailed member documentation.
+# The default value is: NO.
+
+SORT_MEMBERS_CTORS_1ST = NO
```
Comment (KitsuneRal) [inline Doxyfile]:
```suggestion
SORT_MEMBERS_CTORS_1ST = YES
```

File: Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2764 @@
...
+# The EXCLUDE_SYMBOLS tag can be used to specify one or more symbol names
+# (namespaces, classes, functions, etc.) that should be excluded from the
+# output. The symbol name can be a fully qualified name, a word, or if the
+# wildcard * is used, a substring. Examples: ANamespace, AClass,
+# ANamespace::AClass, ANamespace::*Test
+#
+# Note that the wildcards are matched against the file with absolute path, so to
+# exclude all test directories use the pattern */test/*
+
+EXCLUDE_SYMBOLS        = Quotient::_impl::* Quotient::JsonConverter<*> Quotient::JsonObjectConverter<*>
```
Comment (KitsuneRal) [inline Doxyfile]:
```suggestion
EXCLUDE_SYMBOLS        = Quotient::_impl Quotient::JsonConverter<*> Quotient::JsonObjectConverter<*>
```
Also, `_impl::*` somehow didn't seem to work, with thing still showing up under it.

File: Doxyfile [hunk 3]
```diff
@@ -0,0 +1,2764 @@
...
+LATEX_TIMESTAMP        = NO
+
+# The LATEX_EMOJI_DIRECTORY tag is used to specify the (relative or absolute)
+# path from which the emoji images will be read. If a relative path is entered,
+# it will be relative to the LATEX_OUTPUT directory. If left blank the
+# LATEX_OUTPUT directory will be used.
+# This tag requires that the tag GENERATE_LATEX is set to YES.
+
+LATEX_EMOJI_DIRECTORY  =
+
```
Comment (KitsuneRal) [inline Doxyfile]: I don't think we'll ever need this?

File: Doxyfile [hunk 4]
```diff
@@ -0,0 +1,2764 @@
...
+PERLMOD_PRETTY         = YES
+
+# The names of the make variables in the generated doxyrules.make file are
+# prefixed with the string contained in PERLMOD_MAKEVAR_PREFIX. This is useful
+# so different doxyrules.make files included by the same Makefile don't
+# overwrite each other's variables.
+# This tag requires that the tag GENERATE_PERLMOD is set to YES.
+
+PERLMOD_MAKEVAR_PREFIX =
+
```
Comment (KitsuneRal) [inline Doxyfile]: And these.

File: Doxyfile [hunk 5]
```diff
@@ -0,0 +1,2764 @@
...
+DOT_WRAP_THRESHOLD     = 17
+
+# If the TEMPLATE_RELATIONS tag is set to YES then the inheritance and
+# collaboration graphs will show the relations between templates and their
+# instances.
+# The default value is: NO.
+# This tag requires that the tag HAVE_DOT is set to YES.
+
+TEMPLATE_RELATIONS     = NO
```
Comment (KitsuneRal) [inline Doxyfile]:
```suggestion
TEMPLATE_RELATIONS     = YES
```
maybe?

**Review by KitsuneRal — COMMENTED** [review]
I have no prior experience with Doxyfiles but my uneducated suggestions are below :) Also: THANKS!

**Comment by KitsuneRal** [comment]
One more thing I forgot: I wonder if using the Clang parser would help with recognising `[[deprecated]]` declarations that don't have `\deprecated` counterparts in doc-comments.

**Comment by TobiasFella** [comment]
I'm also proposing to use the theme currently deployed at https://quotient.tobiasfella.de/ (not in the merge request yet); opinions?

File: .github/workflows/static.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      url: ${{ steps.deployment.outputs.page_url }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install doxygen
+        run: sudo apt-get -qq install doxygen 
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Build docs
+        run: doxygen
+        shell: bash
```
Comment (KitsuneRal) [inline .github/workflows/static.yml]: Is it actually needed here? Bash is implied by `ubuntu-latest`.

**Review by KitsuneRal — COMMENTED** [review]
Thanks a lot! A nitpick about the shell override.

**Review by KitsuneRal — APPROVED** [review]

**Pull request merged**